accepted/future-releases/abstract-external-fields/feature-specification.md

```diff
@@ -0,0 +1,240 @@
+
+The features specified in this document are syntactic sugar, that is, they
+are specified in terms of a small program transformation that eliminates
+them. This fully determines the further static analysis (including errors
+and warnings), and the dynamic semantics. The transformations are as
+follows:
+
+An abstract instance variable declaration _D_ is treated as an abstract
+getter declaration and possibly an abstract setter declaration. The setter
+is included if and only if _D_ is non-final. The return type of the getter
+and the parameter type of the setter, if present, is the type of _D_
+(*which may be declared explicitly, obtained by override inference, or
+defaulted to `dynamic`*). The parameter of the setter, if present, has the
+modifier `covariant` if and only if _D_ has the modifier `covariant`. _For
+example:_
+
+```dart
+abstract class A {
+  // Abstract instance variables.
+  abstract int i1, i2;
+  abstract var x;
+  abstract final int fi;
+  abstract final fx;
+  abstract covariant num cn;
+  abstract covariant var cx;
+
+  // Desugared meaning of the above.
+  int get i1;
+  void set i1(int _);
+  int get i2;
+  void set i2(int _);
+  dynamic get x;
+  void set x(dynamic _);
+  int get fi;
+  dynamic get fx;
+  num get cn;
+  void cn(covariant num _);
+  dynamic get cx;
+  void set cx(covariant dynamic _);
+}
+```
+
+An external variable declaration _D_ is treated as an external getter
+declaration and possibly an external setter declaration. The setter is
+included if and only if _D_ is non-final. The return type of the getter and
+the parameter type of the setter, if present, is the type of _D_ (*which
+may be declared explicitly, obtained by override inference, or defaulted to
+`dynamic`*). The parameter of the setter, if present, has the modifier
+`covariant` if and only if _D_ has the modifier `covariant` (*the grammar
+only allows this modifier on external instance variables*). _For example:_
+
+```dart
+// External top level variables. PS: Covariance not supported at top level.
+external int i1; // ...
+external final fx;
+
+// Desugared meaning of the above.
+external int get i1;
+external void set i1(int _); // ...
+external dynamic get fx;
+
+class A {
+  // External instance variables.
+  external int i1; // ...
+  external covariant var cx;
+
+  // Desugared meaning of external instance variables.
+  external int get i1;
+  external void set i1(int _); // ...
+  external dynamic get cx;
+  external void set cx(covariant dynamic _);
+
+  // External static variables. PS: Covariance not supported with static.
+  external static int i1; // ...
+  external static final fx;
+
+  // Desugared meaning of external static variables.
+  external static int get i1;
+  external static void set i1(int _); // ...
+  external static dynamic get fx;
+}
+```
+
+Metadata on an abstract or external variable declaration _D_ is associated
+with both the getter and the setter, if present, that arise by desugaring
+of _D_.
```

Comment:
Maybe add ", just as for a concrete variable declaration" (so readers won't think this is special for abstract ones).

Reply:
But that is actually not true. At least, according to 'dart:mirrors' and reflectable, metadata is associated with concrete declarations, in this case the variable declaration, and implicitly induced getters/setters do not have the same metadata.

But with external/abstract variables there is no such variable, so the metadata needs to be on the getter/setter directly. Luckily, canonicalization implies that it is benign to copy the metadata syntax, because that syntax ends up in the same binding environment and will denote the same objects on the accessors as it did on the variable which was eliminated.